memory_core/pond_core.py

```diff
@@ -0,0 +1,431 @@
+                                          io_info.port_name))
+                # TODO - commented out for now. re-visit
+                assert(len(self.__outputs) > 0)
+
+        # We call clk_en stall at this level for legacy reasons????
+        self.add_ports(
+            stall=magma.In(TBit),
+        )
+
+        # put a 1-bit register and a mux to select the control signals
+        for control_signal, width in control_signals:
+            if width == 1:
+                mux = MuxWrapper(2, 1, name=f"{control_signal}_sel")
+                reg_value_name = f"{control_signal}_reg_value"
+                reg_sel_name = f"{control_signal}_reg_sel"
+                self.add_config(reg_value_name, 1)
+                self.add_config(reg_sel_name, 1)
+                self.wire(mux.ports.I[0], self.ports[control_signal])
+                self.wire(mux.ports.I[1], self.registers[reg_value_name].ports.O)
+                self.wire(mux.ports.S, self.registers[reg_sel_name].ports.O)
+                # 0 is the default wire, which takes from the routing network
+                self.wire(mux.ports.O[0], self.underlying.ports[control_signal][0])
+            else:
+                for i in range(width):
+                    mux = MuxWrapper(2, 1, name=f"{control_signal}_{i}_sel")
+                    reg_value_name = f"{control_signal}_{i}_reg_value"
+                    reg_sel_name = f"{control_signal}_{i}_reg_sel"
+                    self.add_config(reg_value_name, 1)
+                    self.add_config(reg_sel_name, 1)
+                    self.wire(mux.ports.I[0], self.ports[f"{control_signal}_{i}"])
+                    self.wire(mux.ports.I[1], self.registers[reg_value_name].ports.O)
+                    self.wire(mux.ports.S, self.registers[reg_sel_name].ports.O)
+                    # 0 is the default wire, which takes from the routing network
+                    self.wire(mux.ports.O[0], self.underlying.ports[control_signal][i])
+
+        # Wire the other signals up...
+        for pname, pdir, expl_arr, ind, uname in other_signals:
+            # If we are in an explicit array moment, use the given wire name...
+            if expl_arr is False:
+                # And if not, use the index
+                self.wire(self.ports[pname][0], self.underlying.ports[uname][ind])
+            else:
+                self.wire(self.ports[pname], self.underlying.ports[pname])
+
+        # CLK, RESET, and STALL PER STANDARD PROCEDURE
+
+        # Need to invert this
+        self.resetInverter = FromMagma(mantle.DefineInvert(1))
+        self.wire(self.resetInverter.ports.I[0], self.ports.reset)
+        self.wire(self.convert(self.resetInverter.ports.O[0],
+                               magma.asyncreset),
+                  self.underlying.ports.rst_n)
+        self.wire(self.ports.clk, self.underlying.ports.clk)
+
+        # Mem core uses clk_en (essentially active low stall)
+        self.stallInverter = FromMagma(mantle.DefineInvert(1))
```

Comment:
A lot of code here seems to be duplicated. Maybe another round of refactoring later? For now I think it's fine.